asyn/devEpics/devAsynXXXArray.cpp

```diff
@@ -0,0 +1,573 @@
+            "%s %s::%s ringSize=%d, len=%d, callback data:",
+            pRecord_->name, driverName, functionName, ringSize_, (int)len);
+        if (ringSize_ == 0) {
+            /* Not using a ring buffer */
+            dbScanLock((dbCommon *)pRecord_);
+            if (len > pRecord_->nelm) len = pRecord_->nelm;
+            if (pasynUser->auxStatus == asynSuccess) {
+                for (i=0; i<(int)len; i++) pData[i] = value[i];
+                pRecord_->nord = (epicsUInt32)len;
+            }
+            pRecord_->time = pasynUser->timestamp;
+            result_.status = (asynStatus) pasynUser->auxStatus;
+            result_.alarmStatus = (epicsAlarmCondition) pasynUser->alarmStatus;
+            result_.alarmSeverity = (epicsAlarmSeverity) pasynUser->alarmSeverity;
+            gotValue_++;
+            dbScanUnlock((dbCommon *)pRecord_);
+            if (isOutput_)
+                scanOnce((dbCommon *)pRecord_);
+            else
+                scanIoRequest(ioScanPvt_);
+        } else {
+            /* Using a ring buffer */
+            ringBufferElement *rp;
+
+            /* If interruptAccept is false we just return.  This prevents more ring pushes than pops.
+             * There will then be nothing in the ring buffer, so the first
+             * read will do a read from the driver, which should be OK. */
+            if (!interruptAccept) return;
+
+            epicsMutexLock(ringBufferLock_);
+            rp = &ringBuffer_[ringHead_];
+            if (len > pRecord_->nelm) len = pRecord_->nelm;
+            rp->len = len;
+            for (i=0; i<(int)len; i++) rp->pValue[i] = value[i];
+            rp->time = pasynUser->timestamp;
+            rp->status = (asynStatus) pasynUser->auxStatus;
+            rp->alarmStatus = (epicsAlarmCondition) pasynUser->alarmStatus;
+            rp->alarmSeverity = (epicsAlarmSeverity) pasynUser->alarmSeverity;
+            ringHead_ = (ringHead_ == ringSize_ - 1) ? 0 : ringHead_ + 1;
+            if (ringHead_ == ringTail_) {
+                /* There was no room in the ring buffer.  In the past we just threw away
+                 * the new value.  However, it is better to remove the oldest value from the
+                 * ring buffer and add the new one.  That way the final value the record receives
+                 * is guaranteed to be the most recent value */
+                ringTail_ = (ringTail_ == ringSize_ - 1) ? 0 : ringTail_ + 1;
+                ringBufferOverflows_++;
+            } else {
+                /* We only need to request the record to process if we added a new
+                 * element to the ring buffer, not if we just replaced an element. */
+                if (isOutput_)
+                    scanOnce((dbCommon *)pRecord_);
+                else
+                    scanIoRequest(ioScanPvt_);
+            }
+            epicsMutexUnlock(ringBufferLock_);
+        }
+    }
+};
+
+extern "C" {
+typedef struct analogDset { /* analog  dset */
```

Comment:
I don't think this `extern "C"` should start so early.

Also, with C++ the `typedef struct x {...} x;` pattern is not necessary.  A plain `struct x {...};` is equivalent.

For that matter, I think that the definitions of `struct analogDset` and `class devAsynXXXArray` should be placed in an [anonymous namespace](https://en.cppreference.com/w/cpp/language/namespace#Unnamed_namespaces) to remove any possibility of a name collision (although I think this is already unlikely).

An anonymous namespace (aka. `namespace { ... }`) is the c++ super-set of a C `static` global, which also [hides](https://en.cppreference.com/w/cpp/language/storage_duration#internal_linkage) some symbols which are defined implicitly.  eg. type and exception unwinding info.

eg. in `namespace { static int myglobal; }` the `static` qualifier is redundant.

One relevant example, where [all but](https://github.com/mdavidsaver/pvxs/blob/f0de452930235ae58ee0bf059a816a222b82a9fc/ioc/iochooks.cpp#L25-L243) the [last two](https://github.com/mdavidsaver/pvxs/blob/f0de452930235ae58ee0bf059a816a222b82a9fc/ioc/iochooks.cpp#L245-L261) symbols are hidden.

The overall structure of `devAsynXXXArray.cpp` would then be

```c++
// #include s
namespace {
template<...> class devAsynXXXArray { ... };
struct analogDset { ... };
} // end of namespace
#define MAKE_DEVSUP(...) ...
extern "C" {
MAKE_DEVSUP(...); ...
/// MAKE_DEVSUP expands to only static globals or epicsExport*()
} // end of extern "C"
```

With this structure, running `nm --defined-only --extern-only devAsynXXXArray.o` should list only `pvar_*` symbols as strong definitions (`D`).  The internal symbols will either be (`d` or `t`), or optimized away entirely.  (`man nm` lists all of the letter codes)

Reply:
+1
That's exactly how I am structuring/declaring C++ device support nowadays.